tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,182 @@
+        obs_maybe = await self.store.get_partial_values(
+            prototype=default_buffer_prototype(), key_ranges=key_ranges
+        )
+        return obs_maybe
+
+    async def delete(self, path):  # path is key
+        await self.store.delete(path)
+
+    async def empty(self):
+        await self.store.empty()
+
+    async def clear(self):
+        await self.store.clear()
+
+    async def exists(self, key):
+        raise NotImplementedError
+
+    async def list_dir(self, prefix):
+        raise NotImplementedError
+
+    async def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    async def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    async def supports_listing(self):
+        raise NotImplementedError
+
+    async def supports_partial_writes(self):
+        raise NotImplementedError
+
+    async def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    def __init__(self):  # look into using run_machine_as_test()
+        super().__init__()
+        self.model = {}
+        self.store = SyncStoreWrapper(MemoryStore(mode="w"))
+        # self.store = MemoryStore(mode='w')
+
+    @rule(key=paths, data=st.binary(min_size=0, max_size=100))
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        self.store.set(key, data_buf))
+        self.model[key] = data_buf  # this was data
+
+    @invariant()
+    def check_paths_equal(self) -> None:
+        note("Checking that paths are equal")
+        paths = self.store.list()
+        assert list(self.model.keys()) == paths
+        note("Checking values equal")
+        for key, _val in self.model.items():
+            store_item = self.store.get(key).to_bytes()
+            # note(f'(inv) model item: {self.model[key]}')
+            # note(f'(inv) {store_item=}')
+            assert self.model[key].to_bytes() == store_item
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    # @rule(key=keys_bundle)
+    def get(self, data) -> None:
+        key = data.draw(st.sampled_from(sorted(self.model.keys())))
+        store_value = asyncio.run(self.sync_wrapper.get(key))
+        # to bytes here necessary (on model and store) because data_buf set to model in set()
+        assert self.model[key].to_bytes() == store_value.to_bytes()
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(data=st.data())
+    def get_partial_values(self, data) -> None:
+        key_st = st.sampled_from(sorted(self.model.keys()))
+        key_range = data.draw(key_ranges(keys=key_st))
+
+        obs_maybe = asyncio.run(self.sync_wrapper.get_partial_values(key_range))
+        observed = []
+
+        for obs in obs_maybe:
+            assert obs is not None
```

Comment:
assert that None's are in the same place for the model and the store.

Reply:
didn't see this earlier but addressed [here](https://github.com/zarr-developers/zarr-python/pull/2070/commits/8e6f03c02c920e4979d70fb154dbb2a33f4ca1ee), I think i did what you meant here but not totally sure...

Reply:
now I'm confused. Does it only return `None` for non-existent keys? Does it return `None` for invalid key ranges too.

if the former, that path isn't tested here since you're drawing from the keys bundle.
If the latter, then we'll need to assert those `None`s are in the right place using the model.

Reply:
I added a `get_invalid_keys()` rule to cover this